src/toil/leader.py

```diff
@@ -18,162 +18,507 @@
 
 from toil import resolveEntryPoint
 from toil.jobStores.abstractJobStore import NoSuchJobException
-from toil.jobGraph import JobNode
-from toil.lib.bioio import getTotalCpuTime, logStream
 from toil.provisioners.clusterScaler import ClusterScaler
-
+from toil.serviceManager import ServiceManager
+from toil.statsAndLogging import StatsAndLogging
+from toil.jobGraph import JobNode
+from toil.toilState import ToilState
 
 logger = logging.getLogger( __name__ )
 
-
-####################################################
-##Stats/logging aggregation
 ####################################################
+# Exception thrown by the Leader class when one or more jobs fails
+#################################################### 
 
+class FailedJobsException( Exception ):
+    def __init__(self, jobStoreLocator, failedJobs, jobStore):
+        msg = "The job store '%s' contains %i failed jobs" % (jobStoreLocator, len(failedJobs))
+        try:
+            msg += ": %s" % ", ".join((str(failedJob) for failedJob in failedJobs))
+            for jobNode in failedJobs:
+                job = jobStore.load(jobNode.jobStoreID)
+                if job.logJobStoreFileID:
+                    msg += "\n=========> Failed job %s %s\n" % (jobNode, jobNode.jobStoreID)
+                    with job.getLogFileHandle(jobStore) as fH:
+                        msg += fH.read()
+                    msg += "<=========\n"
+        # catch failures to prepare more complex details and only return the basics
+        except:
+            logger.exception('Exception when compiling information about failed jobs')
+        super( FailedJobsException, self ).__init__(msg)
+        self.jobStoreLocator = jobStoreLocator
+        self.numberOfFailedJobs = len(failedJobs)
+        
+####################################################
+# Exception thrown by the Leader class when a deadlock is encountered due to insufficient
+# resources to run the workflow
+#################################################### 
 
-class StatsAndLogging( object ):
-    """
-    Class manages a thread that aggregates statistics and logging information on a toil run.
-    """
+class DeadlockException( Exception ):
+    def __init__(self, msg):
+        msg = "Deadlock encountered: " + msg
+        super( DeadlockException, self ).__init__(msg)
 
-    def __init__(self, jobStore):
-        # Start the stats/logging aggregation thread
-        self._stop = Event()
-        self._worker = Thread(target=self.statsAndLoggingAggregator,
-                              args=(jobStore, self._stop))
-        self._worker.start()
+####################################################
+##Following class represents the leader
+####################################################
 
-    @staticmethod
-    def statsAndLoggingAggregator(jobStore, stop):
+# Represents a job and its requirements as issued to the batch system
+IssuedJob = namedtuple("IssuedJob", "jobStoreID memory cores disk preemptable")
```

Comment:
This isn't used anywhere, I think it was removed in the job name PR. Maybe reintroduced in a merge error?